packet/src/pacer.rs

```diff
@@ -0,0 +1,1032 @@
+
+        let total_packet_count: u32 = self.queue_states.iter().map(|q| q.packet_count).sum();
+        if now < next_send_time {
+            if total_packet_count > 0 {
+                let diff = next_send_time.saturating_duration_since(now);
+                debug!(?now, ?diff, ?next_send_time, "Delaying send");
+            }
+
+            if queue.is_some() {
+                // We had a queue to send on
+                self.next_send_time = Some(next_send_time);
+            }
+            self.next_poll_outcome = None;
+            return;
+        }
+
+        if let Some(queue) = queue {
+            self.next_poll_outcome = Some(PollOutcome::PollQueue(queue.id));
+            self.next_send_time = Some(next_send_time);
+        }
+    }
+
+    fn poll_action(&mut self) -> PollOutcome {
+        let Some(next) = self.next_poll_outcome.take() else {
+            return PollOutcome::Nothing;
+        };
+
+        self.need_immediate_timeout = true;
+        next
+    }
+
+    fn register_send(&mut self, now: Instant, packet_size: DataSize, _from: QueueId) {
+        self.media_debt += packet_size;
+        self.media_debt = self
+            .media_debt
+            .min(self.adjusted_bitrate * MAX_MEDIA_DEBT_IN_TIME);
+        self.last_send_time = Some(now);
+    }
+}
+
+impl LeakyBucketPacer {
+    pub fn new(target_bitrate: Bitrate, pacing: Duration) -> Self {
+        const DEFAULT_QUEUE_LIMIT: Duration = Duration::from_secs(2);
+
+        Self {
+            target_bitrate,
+            adjusted_bitrate: Bitrate::ZERO,
+            last_handle_time: None,
+            last_send_time: None,
+            next_send_time: None,
+            media_debt: DataSize::ZERO,
+            pacing,
+            queue_limit: DEFAULT_QUEUE_LIMIT,
+            queue_states: vec![],
+            need_immediate_timeout: false,
+            next_poll_outcome: None,
+        }
+    }
+
+    fn update_handle_time_and_get_elapsed(&mut self, now: Instant) -> Duration {
+        // Due the calling code this also happens when a packet is queued in any upstream queue.
+        let Some(previous_handle_time) = self.last_handle_time else {
+            self.last_handle_time = Some(now);
+            return Duration::ZERO;
+        };
+
+        let elapsed = now - previous_handle_time;
+        self.last_handle_time = Some(now);
+
+        elapsed
+    }
+
+    fn clear_debt(&mut self, elapsed: Duration) {
+        self.media_debt = self
+            .media_debt
+            .saturating_sub(self.media_debt.min(self.adjusted_bitrate * elapsed));
+    }
+
+    fn next_send_time(&self, now: Instant) -> (Instant, Option<&QueueState>) {
+        // If we have never sent before, do so immediately on an arbitrary non-empty queue.
+        if self.last_send_time.is_none() {
+            let mut queues = self.queue_states.iter().filter(|q| q.packet_count > 0);
+
+            return (now, queues.next());
+        };
+
+        let unpaced_audio = self
```

Comment:
If the purpose of the PacketKind is to pace or not pace, why call it Audio and Video?  Why not instead call it Unpaced and Paced?